### PR TITLE
fix non contiguous tensor value error in save_pretrained

### PR DESCRIPTION
**Pull request by congcongke**
# What does this PR do?
Fixes # [28293](https://github.com/huggingface/transformers/issues/28293)

This issue is disappeared after this fix.
Any way, only classification case is verified. This PR is just a suggestion.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

 @ArthurZucker @muellerzr and @SunMarc

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
Look good to me, would just want a small reproducible snippet please!

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -2742,7 +2742,10 @@ def save_pretrained(
         if module_map:
             filename_to_tensors = logging.tqdm(filename_to_tensors, desc="Saving checkpoint shards")
         for shard_file, tensors in filename_to_tensors:
-            shard = {tensor: state_dict[tensor] for tensor in tensors}
+            shard = {
+                tensor: state_dict[tensor] if state_dict[tensor].is_contiguous() else state_dict[tensor].contiguous()
```
Comment (amyeroberts) [inline src/transformers/modeling_utils.py:2746]: I don't think you need to check `is_contiguous` here - calling `contiguous` will just return the instance if it's already contiguous

**Comment by congcongke** [comment]
> Look good to me, would just want a small reproducible snippet please!

reproduce keys:

- dataset: https://huggingface.co/datasets/dirtycomputer/simplifyweibo_4_moods/tree/main
- model: https://huggingface.co/hfl/chinese-roberta-wwm-ext/tree/main
- Hardware: Tesla4 * 2
- Framework: PytorchDDP

`
model = AutoModelForSequenceClassification.from_pretrained(args.model_path, num_labels=4)

trainingArg = TrainingArguments(
        output_dir=args.output_path,
        learning_rate=2e-5,
        per_device_train_batch_size=args.batch_size,
        per_device_eval_batch_size=args.batch_size,
        num_train_epochs=args.epoch,
        weight_decay=0.01,
        evaluation_strategy="epoch",
        save_strategy="epoch",
        load_best_model_at_end=True,
        disable_tqdm=True,
        ddp_backend="nccl",
        ddp_timeout=args.ddp_timeout,
        save_on_each_node=False,
        seed=random.randint(0, 100000),
        push_to_hub=False)

    trainer = Trainer(
        model=model,
        args=trainingArg,
        train_dataset=tokenized_imdb["train"],
        eval_dataset=tokenized_imdb["train"],
        tokenizer=tokenizer,
        data_collator=data_collator,
        compute_metrics=compute_metrics,
    )

    trainer.train()
`

File: src/transformers/modeling_utils.py (hunk 1 above)
Comment (congcongke) [reply src/transformers/modeling_utils.py]: All right. More clever. Thanks!

**Comment by ZhanGHanG9991** [comment]
I encounter this problem today when I fine-tune google-t5/t5-base model and try to save the fine-tuned model.

System Info:
Transformers version: 4.43.4
pytorch version: 2.4.0
Python version: 3.8.19

image

**Comment by congcongke** [comment]
> I encounter this problem today when I fine-tune google-t5/t5-base model and try to save the fine-tuned model.
> 
> System Info: Transformers version: 4.43.4 pytorch version: 2.4.0 Python version: 3.8.19
> 
> image

Would you take this PR as a trial please.

**Review by amyeroberts — APPROVED** [review]
Thanks!

**Comment by amyeroberts** [comment]
@ZhanGHanG9991 If you can confirm this fixes your error I think we're good to merge!

**Review by muellerzr — APPROVED** [review]
agreed this solution looks good!

**Review by SunMarc — APPROVED** [review]
Nice !

**Review by PierrickPochelu — APPROVED** [review]

**Comment by amyeroberts** [comment]
Let's merge as there's many approvals and could be blocking

**Comment by ZhanGHanG9991** [comment]
> @ZhanGHanG9991 If you can confirm this fixes your error I think we're good to merge!

Sorry for not replying in time. I just tried it and the bug has been solved after modification. No more errors!

**Comment by ZhanGHanG9991** [comment]
> > I encounter this problem today when I fine-tune google-t5/t5-base model and try to save the fine-tuned model.
> > System Info: Transformers version: 4.43.4 pytorch version: 2.4.0 Python version: 3.8.19
> > image
> 
> Would you take this PR as a trial please.

I've tried it and the bug has been solved after modification! Thank you!
